tempesta_fw/tls.c

```diff
@@ -379,23 +388,87 @@ tfw_tls_encrypt(struct sock *sk, struct sk_buff *skb, unsigned int limit)
 
 	if (likely(sgt.nents <= AUTO_SEGS_N)) {
 		sgt.sgl = sg;
+		out_sgt.sgl = out_sg;
+		pages = auto_pages;
 	} else {
 		sgt.sgl = kmalloc(sizeof(struct scatterlist) * sgt.nents,
 				  GFP_ATOMIC);
 		if (!sgt.sgl) {
 			T_WARN("cannot alloc TLS encryption scatter list.\n");
 			return -ENOMEM;
 		}
+		out_sgt.sgl = kmalloc(sizeof(struct scatterlist) * out_sgt.nents,
+				      GFP_ATOMIC);
+		if (!out_sgt.sgl) {
+			T_WARN("cannot alloc TLS encryption scatter list.\n");
+			return -ENOMEM;
+		}
+		pages = kmalloc(sizeof(struct page *) * out_sgt.nents,
+				GFP_ATOMIC);
+		if (!pages) {
+			T_WARN("cannot alloc memory.\n");
+			return -ENOMEM;
+		}
 	}
 	sg_init_table(sgt.sgl, sgt.nents);
+	sg_init_table(out_sgt.sgl, out_sgt.nents);
 
-	for (next = skb, frags = 0; ; ) {
+	remain = offset = 0;
+	for (next = skb, frags = 0, out_frags = 0, elt = 0; ; ) {
 		/*
 		 * skb data and tails are already adjusted above,
 		 * so use zero offset and skb->len.
 		 */
 		r = skb_to_sgvec(next, sgt.sgl + frags, 0, next->len);
 
+		if (skb_shinfo(next)->tx_flags & SKBTX_SHARED_FRAG ||
+		    tls->resp_from_cache)
```

Comment:
Just a thought. All skb's with shared should have fragments point to pages which are also held elsewhere. So can we determine which fragments contain shared data by checking `page_ref_count(skb_frag_page(f))`?

Reply:
Also, how about setting `SKBTX_SHARED_FRAG` flag in `tfw_cache_build_resp_body()` where generated skb's with data from the cache? If that works, it may make passing `resp_from_cache` flag through the call chain obsolete.

Reply:
Agree here. The whole `resp_from_cache` thing looks like messy workaround. Anyway, the copy should happen not for the whole message, but only for the body. If setting `SKBTX_SHARED_FRAG` flag is not applicable for the skb (some very specific processing insed the kernel), we can copy data in `cache.c` and use in-place crypto operations later.

Reply:
@ikoveshnikov, how do I know where the body is the message, and where are the headers in the message? In my opinion, if there is a flag SKBTX_SHARED_FRAG you need to copy.

Reply:
All answers in Rinat's comments.

Reply:
But why do you care where the headers or the body are? Isn't it enough to mark body data skb's coming from the cache as shared?

Reply:
>So can we determine which fragments contain shared data by checking page_ref_count(skb_frag_page(f))?

I first thought that the answer to the question is yes. But the answer is no

Reply:
> we can copy cached response body
This is will copy, but we wish w/o copy.